File: leet_BTREE_236_lowest_common_ancestor.scala
/**
 * Definition for a binary tree node.
 * class TreeNode(var _value: Int) {
 *   var value: Int = _value
 *   var left: TreeNode = null
 *   var right: TreeNode = null
 * }
 */

object Solution {
    def lowestCommonAncestor(node: TreeNode, p: TreeNode, q: TreeNode): TreeNode = 
    {
        if (node == null || node == p || node == q) node
        else
        {
            val left = lowestCommonAncestor(node.left, p, q)
            val right = lowestCommonAncestor(node.right, p, q)
            if (left != null && right != null) node
            else if (left != null) left
            else right
        }
    }
}

File: leet_BTREE_236_lowest_common_ancestor.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, node: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        if not node or node == p or node == q:
            return node
        left = self.lowestCommonAncestor(node.left, p, q)
        right = self.lowestCommonAncestor(node.right, p, q)
        if not left:
            return right
        if not right:
            return left
        return node
